modules/popup-stack/lib/PopupStack.ts

```diff
@@ -0,0 +1,220 @@
+   * page. Z-indexes are set when the item is added to the stack. If your application requires
+   * popups to be registered initially, but rendered when the user triggers some event, call this
+   * method when the event triggers.
+   */
+  add(item: PopupStackItem): void {
+    stack.items.push(item);
+
+    setZIndexOfElements(PopupStack.getElements());
+  },
+
+  /**
+   * Removes an item from the stack by its `HTMLElement` reference. This should be called when a
+   * popup is "closed" or when the element is removed from the page entirely to ensure proper memory
+   * cleanup. This will not automatically be called when the element is removed from the DOM. Will
+   * reset z-index values of the stack
+   */
+  remove(element: HTMLElement): void {
+    stack.items = stack.items.filter(item => item.element !== element);
+
+    setZIndexOfElements(PopupStack.getElements());
+  },
+
+  /**
+   * Returns true when the provided `element` is at the top of the stack. It will return false if it
+   * is not the top of the stack or is not found in the stack. The `element` should be the same
+   * reference that was passed to `add`
+   */
+  isTopmost(element: HTMLElement): boolean {
+    const last = getLast(stack.items);
+
+    if (last) {
+      return last.element === element;
+    }
+    return false;
+  },
+
+  /**
+   * Returns an array of elements defined by the `element` passed to `add`.
+   */
+  getElements(): HTMLElement[] {
+    return stack.items.map(i => i.element);
+  },
+
+  /**
+   * Bring the element to the top of the stack. This is useful for persistent popups to place them
+   * on top of the stack when clicked. If an `owner` was provided to an item when it was added and
+   * that owner is a DOM child of another item in the stack, that item will be considered a "parent"
+   * to this item. If the previous are true, all "children" stack items will be brought to top as
+   * well and will be on top of the element passed to `bringToTop`. This maintains stack item
+   * "hierarchy" so that stack items like Popups and Tooltips don't get pushed behind elements they
+   * are supposed to be on top of.
+   *
+   * This does not need to be called when a popup is added since added popups are already place on
+   * the top of the stack.
+   */
+  bringToTop(element: HTMLElement): void {
+    const item = stack.items.find(i => i.element === element);
+
+    if (item) {
+      stack.items = [...stack.items.filter(i => i !== item), item];
+
+      // Also bring children to top. There are a few cases where stacking might break otherwise:
+      // - Clicking a Popup calls `bringToTop`, but mouse is over a Tooltip so that Tooltip is now
+      //   under the Popup
+      // - Clicking a button opens a new Popup, but that click bubbles up to a `bringToTop` call
+      //   putting the new popup under an existing one
+      getChildPopups(item, stack.items).forEach(popup => {
+        PopupStack.bringToTop(popup.element);
+      });
```

Comment:
If we didn't do this, there could be a race condition that would cause the new popup to appear behind a window with `bringToFront`:

![popup_dom_event_bubbling](https://user-images.githubusercontent.com/338257/83924476-031af580-a742-11ea-8f68-0edabdf0fd6a.gif)